File: lintcode/python/0595_binary_tree_longest_consecutive_sequence.py
#!/usr/bin/python -t

# binary tree
# divid and conquer way
# time O(n)
# space O(n)


class Solution:
    """
    @param root: the root of binary tree
    @return: the length of the longest consecutive sequence path
    """
    def longest_consecutive(self, root: TreeNode) -> int:
        # write your code here
        if not root:
            return 0

        ret, _ = self.helper(root)

        return ret

    def helper(self, node):
        if not node:
            return 0, 0

        left, left_ret = self.helper(node.left)
        right, right_ret = self.helper(node.right)

        cur_ret = 1
        if node.left and node.val == node.left.val - 1:
            cur_ret = max(left_ret+1, cur_ret)

        if node.right and node.val == node.right.val - 1:
            cur_ret = max(right_ret+1, cur_ret)

        ret = max(cur_ret, left, right)

        return ret, cur_ret



# binary tree
# traversal and divid and conquer way
# time O(n)
# space O(n)

class Solution:
    def __init__(self):
        self.ret = 0

    """
    @param root: the root of binary tree
    @return: the length of the longest consecutive sequence path
    """
    def longest_consecutive(self, root: TreeNode) -> int:
        # write your code here
        if not root:
            return 0

        self.helper(root)

        return self.ret

    def helper(self, node):
        if not node:
            return 0

        left = self.helper(node.left)
        right = self.helper(node.right)

        ret = 1
        if node.left and node.val == node.left.val - 1:
            ret = max(left+1, ret)

        if node.right and node.val == node.right.val - 1:
            ret = max(right+1, ret)

        self.ret = max(self.ret, ret)

        return ret

# dfs, traversal and divid and conquer solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of binary tree
    @return: the length of the longest consecutive sequence path
    """
    def longestConsecutive(self, root):
        # write your code here
        return self.helper(root, None, 0)
        
    def helper(self, node, parent, length):
        if node == None:
            return length
            
        if parent and node.val == parent.val+1:
            length += 1
        else:
            length = 1
            
        return max(length, self.helper(node.left, node, length),
                            self.helper(node.right, node, length))

File: lintcode/golang/0595_binary_tree_longest_consecutive_sequence.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the root of binary tree
 * @return: the length of the longest consecutive sequence path
 */

import "fmt"

func longestConsecutive (root *TreeNode) int {
    // write your code here
    if root == nil {
        return 0
    }
    
    var ret int
    _ = helper(root, &ret)
    
    return ret
}

func helper(root *TreeNode, maxVal *int) int {
    if root == nil {
        return 0
    }
    
    if root.Left == nil && root.Right == nil {
        return 1
    }
    
    left := helper(root.Left, maxVal)
    right := helper(root.Right, maxVal)
    
    fmt.Println(root.Val, left, right)
    var ret int
    ret = max(left, right)
    if root.Left != nil && root.Val + 1 == root.Left.Val {
        fmt.Println("left", root.Left.Val, root.Val, left, ret)
        ret = max(ret, left+1)
    } else {
        left = 0   
    }
    
    if root.Right != nil && root.Val +1 == root.Right.Val {
        fmt.Println("right", root.Right.Val, root.Val, right, ret)
        ret = max(ret, right+1)
    } else {
        right = 0
    }
    
    *maxVal = max(*maxVal, ret)
    if left == 0 && right == 0 {
        ret = 1
    }
    
    return ret
}

func max (a, b int) int {
    if a > b {
        return a
    }
    return b
}
